src/compiler/scala/tools/nsc/typechecker/Duplicators.scala

```diff
@@ -229,7 +229,12 @@ abstract class Duplicators extends Analyzer {
 
         case ddef @ DefDef(_, _, _, _, tpt, rhs) =>
           ddef.tpt modifyType fixType
-          super.typed(ddef.clearType(), mode, pt)
+          val result = super.typed(ddef.clearType(), mode, pt)
+          // TODO this is a hack, we really need a cleaner way to transport symbol attachments to duplicated methods
+          // bodies in specialized subclasses.
+          if (ddef.hasAttachment[DelambdafyTarget.type])
+            result.symbol.updateAttachment(DelambdafyTarget)
+          result
```

Comment:
If we compile:

```
class C[@specialized A] {
  def foo(a: A) = { def bar = 42; a }
}

```

We get to:

https://github.com/scala/scala/blob/92affd3e76d1a3019f1c35ac20af32360d914641/src/compiler/scala/tools/nsc/typechecker/Duplicators.scala#L167-L167

For:

```
tree: def bar(): Int = 42
```

From a call stack:

```
"main@1" prio=5 tid=0x1 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	  at scala.tools.nsc.typechecker.Duplicators$BodyDuplicator.invalidate(Duplicators.scala:167)
	  at scala.tools.nsc.typechecker.Duplicators$BodyDuplicator.scala$tools$nsc$typechecker$Duplicators$BodyDuplicator$$$anonfun$18(Duplicators.scala:181)
	  at scala.tools.nsc.typechecker.Duplicators$BodyDuplicator.scala$tools$nsc$typechecker$Duplicators$BodyDuplicator$$$anonfun$18$adapted(Duplicators.scala:181)
	  at scala.tools.nsc.typechecker.Duplicators$BodyDuplicator$$Lambda$991.478209179.apply(Unknown Source:-1)
	  at scala.collection.immutable.List.foreach(List.scala:376)
	  at scala.tools.nsc.typechecker.Duplicators$BodyDuplicator.invalidateAll(Duplicators.scala:181)
	  at scala.tools.nsc.typechecker.Duplicators$BodyDuplicator.typed(Duplicators.scala:220)
	  at scala.tools.nsc.typechecker.Typers$Typer.transformedOrTyped(Typers.scala:5602)
	  at scala.tools.nsc.typechecker.Typers$Typer.typedDefDef(Typers.scala:2243)
	  at scala.tools.nsc.typechecker.Typers$Typer.typedMemberDef$1(Typers.scala:5305)
	  at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5356)
	  at scala.tools.nsc.typechecker.Typers$Typer.runTyper$1(Typers.scala:5393)
	  at scala.tools.nsc.typechecker.Typers$Typer.typedInternal(Typers.scala:5423)
	  at scala.tools.nsc.typechecker.Typers$Typer.body$2(Typers.scala:5367)
	  at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5371)
	  at scala.tools.nsc.typechecker.Duplicators$BodyDuplicator.typed(Duplicators.scala:232)
	  at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5448)
	  at scala.tools.nsc.typechecker.Duplicators.retyped(Duplicators.scala:36)
	  at scala.tools.nsc.transform.SpecializeTypes$SpecializationDuplicator.super$retyped(SpecializeTypes.scala:1331)
	  at scala.tools.nsc.transform.SpecializeTypes$SpecializationDuplicator.scala$tools$nsc$transform$SpecializeTypes$SpecializationDuplicator$$$anonfun$154(SpecializeTypes.scala:1331)
	  at scala.tools.nsc.transform.SpecializeTypes$SpecializationDuplicator$$Lambda$963.1043639521.apply(Unknown Source:-1)
	  at scala.reflect.internal.SymbolTable.enteringPhase(SymbolTable.scala:235)
	  at scala.tools.nsc.Global.enteringSpecialize(Global.scala:967)
	  at scala.tools.nsc.transform.SpecializeTypes$SpecializationDuplicator.retyped(SpecializeTypes.scala:1331)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.duplicateBody(SpecializeTypes.scala:1791)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.scala$tools$nsc$transform$SpecializeTypes$$anon$2$$$anonfun$196(SpecializeTypes.scala:1641)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2$$Lambda$950.1527007086.apply(Unknown Source:-1)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.reportError(SpecializeTypes.scala:1445)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transformDefDef$1(SpecializeTypes.scala:1641)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transform1(SpecializeTypes.scala:1713)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.scala$tools$nsc$transform$SpecializeTypes$$anon$2$$$anonfun$164(SpecializeTypes.scala:1453)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2$$Lambda$816.778720569.apply(Unknown Source:-1)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.reportError(SpecializeTypes.scala:1445)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transform(SpecializeTypes.scala:1453)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transform(SpecializeTypes.scala:1396)
	  at scala.reflect.api.Trees$Transformer.scala$reflect$api$Trees$Transformer$$$anonfun$6(Trees.scala:2559)
	  at scala.reflect.api.Trees$Transformer$$Lambda$703.2052435819.apply(Unknown Source:-1)
	  at scala.collection.immutable.List.loop$1(List.scala:171)
	  at scala.collection.immutable.List.mapConserve(List.scala:187)
	  at scala.reflect.api.Trees$Transformer.transformTrees(Trees.scala:2559)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.scala$tools$nsc$transform$SpecializeTypes$$anon$2$$$anonfun$190(SpecializeTypes.scala:1624)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2$$Lambda$911.1692073709.apply(Unknown Source:-1)
	  at scala.reflect.api.Trees$Transformer.atOwner(Trees.scala:2600)
	  at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:30)
	  at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:25)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transformTemplate$1(SpecializeTypes.scala:1624)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transform1(SpecializeTypes.scala:1626)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.scala$tools$nsc$transform$SpecializeTypes$$anon$2$$$anonfun$164(SpecializeTypes.scala:1453)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2$$Lambda$816.778720569.apply(Unknown Source:-1)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.reportError(SpecializeTypes.scala:1445)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transform(SpecializeTypes.scala:1453)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transform(SpecializeTypes.scala:1396)
	  at scala.reflect.api.Trees$Transformer.transformTemplate(Trees.scala:2563)
	  at scala.reflect.internal.Trees.scala$reflect$internal$Trees$$$anonfun$37(Trees.scala:1412)
	  at scala.reflect.internal.Trees$$Lambda$699.2089589253.apply(Unknown Source:-1)
	  at scala.reflect.api.Trees$Transformer.atOwner(Trees.scala:2600)
	  at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:30)
	  at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:25)
	  at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:17)
	  at scala.reflect.internal.Trees.itransform(Trees.scala:1411)
	  at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	  at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	  at scala.reflect.api.Trees$Transformer.transform(Trees.scala:2555)
	  at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.transform(TypingTransformers.scala:44)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transform1(SpecializeTypes.scala:1737)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.scala$tools$nsc$transform$SpecializeTypes$$anon$2$$$anonfun$164(SpecializeTypes.scala:1453)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2$$Lambda$816.778720569.apply(Unknown Source:-1)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.reportError(SpecializeTypes.scala:1445)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transform(SpecializeTypes.scala:1453)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transform(SpecializeTypes.scala:1396)
	  at scala.reflect.api.Trees$Transformer.scala$reflect$api$Trees$Transformer$$$anonfun$13(Trees.scala:2589)
	  at scala.reflect.api.Trees$Transformer$$Lambda$698.1611370719.apply(Unknown Source:-1)
	  at scala.collection.immutable.List.loop$1(List.scala:171)
	  at scala.collection.immutable.List.mapConserve(List.scala:187)
	  at scala.reflect.api.Trees$Transformer.transformStats(Trees.scala:2587)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.scala$tools$nsc$transform$SpecializeTypes$$anon$2$$$anonfun$186(SpecializeTypes.scala:1610)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2$$Lambda$818.41765385.apply(Unknown Source:-1)
	  at scala.reflect.api.Trees$Transformer.atOwner(Trees.scala:2600)
	  at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:30)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transform1(SpecializeTypes.scala:1608)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.scala$tools$nsc$transform$SpecializeTypes$$anon$2$$$anonfun$164(SpecializeTypes.scala:1453)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2$$Lambda$816.778720569.apply(Unknown Source:-1)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.reportError(SpecializeTypes.scala:1445)
	  at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transform(SpecializeTypes.scala:1453)
	  at scala.tools.nsc.transform.SpecializeTypes$SpecializationTransformer.scala$tools$nsc$transform$SpecializeTypes$SpecializationTransformer$$$anonfun$269(SpecializeTypes.scala:1983)
	  at scala.tools.nsc.transform.SpecializeTypes$SpecializationTransformer$$Lambda$815.1415630650.apply(Unknown Source:-1)
	  at scala.reflect.internal.SymbolTable.enteringPhase(SymbolTable.scala:235)
	  at scala.reflect.internal.SymbolTable.exitingPhase(SymbolTable.scala:256)
	  at scala.tools.nsc.Global.exitingSpecialize(Global.scala:957)
	  at scala.tools.nsc.transform.SpecializeTypes$SpecializationTransformer.transform(SpecializeTypes.scala:1983)
	  at scala.tools.nsc.transform.SpecializeTypes$SpecializationTransformer.transform(SpecializeTypes.scala:1979)
	  at scala.tools.nsc.ast.Trees$Transformer.transformUnit(Trees.scala:140)
	  at scala.tools.nsc.transform.Transform$Phase.apply(Transform.scala:30)
	  at scala.tools.nsc.Global$GlobalPhase.scala$tools$nsc$Global$GlobalPhase$$$anonfun$11(Global.scala:402)
	  at scala.tools.nsc.Global$GlobalPhase$$Lambda$387.888655833.apply$mcV$sp(Unknown Source:-1)
	  at scala.tools.nsc.Global$GlobalPhase.withCurrentUnitNoLog(Global.scala:395)
	  at scala.tools.nsc.Global$GlobalPhase.withCurrentUnit(Global.scala:386)
	  at scala.tools.nsc.Global$GlobalPhase.applyPhase(Global.scala:402)
	  at scala.tools.nsc.Global$GlobalPhase.scala$tools$nsc$Global$GlobalPhase$$$anonfun$10(Global.scala:366)
	  at scala.tools.nsc.Global$GlobalPhase.scala$tools$nsc$Global$GlobalPhase$$$anonfun$10$adapted(Global.scala:366)
	  at scala.tools.nsc.Global$GlobalPhase$$Lambda$386.980138431.apply(Unknown Source:-1)
	  at scala.collection.Iterator.foreach(Iterator.scala:932)
	  at scala.collection.AbstractIterator.foreach(Iterator.scala:1409)
	  at scala.tools.nsc.Global$GlobalPhase.run(Global.scala:366)
	  at scala.tools.nsc.Global$Run.compileUnitsInternal(Global.scala:1396)
	  at scala.tools.nsc.Global$Run.compileUnits(Global.scala:1381)
	  at scala.tools.nsc.Global$Run.compileSources(Global.scala:1376)
	  at scala.tools.nsc.Global$Run.compile(Global.scala:1470)
	  at scala.tools.nsc.Driver.doCompile(Driver.scala:35)
	  at scala.tools.nsc.MainClass.doCompile(Main.scala:24)
	  at scala.tools.nsc.Driver.process(Driver.scala:55)
	  at scala.tools.nsc.Driver.main(Driver.scala:68)
	  at scala.tools.nsc.Main.main(Main.scala:-1)

```

Note that the delambdafy target method that is still method-local at the specialize phase.